sycl/include/CL/sycl/detail/pi.hpp

```diff
@@ -80,13 +80,72 @@ namespace pi {
     template<typename Exception>
     void check(PiResult Result);
   };
+
+  // The run-time tracing of PI calls.
+  // TODO: replace PiCall completely with this one (PiTrace)
+  //
+  template <typename T> inline
+  void Print(T val) {
+    std::cout << "<unknown> : " << val;
+  }
+
+  template<> inline void Print<> (PiPlatform val) { std::cout << "pi_platform : " << val; }
+  template<> inline void Print<> (PiResult val) {
+    std::cout << "pi_result : ";
+    if (val == PI_SUCCESS)
+      std::cout << "PI_SUCCESS";
+    else
+      std::cout << val; 
+  }
+  
+  inline void PrintArgs(void) {}
+  template <typename Arg0, typename... Args>
+  void PrintArgs(Arg0 arg0, Args... args) {
+    std::cout << std::endl << "       ";
+    Print(arg0);
+    PrintArgs(std::forward<Args>(args)...);
+  }
+  
+  template <typename FnType>
+  class Trace {
+  private:
+    FnType m_FnPtr;
+    static bool m_TraceEnabled;
+  public:
+    Trace(FnType FnPtr, const std::string &FnName) : m_FnPtr(FnPtr) {
+      if (m_TraceEnabled)
+        std::cout << "---> " << FnName << "(";
+    }
+  
+    template <typename... Args>
+    typename std::result_of<FnType(Args...)>::type
+    operator() (Args... args) {
+      if (m_TraceEnabled)
```

Comment:
It sounds weird that methods of class **Trace** traces **optionally**.
Why do you need to introduce new class, can't we add this functionality to PiCall ?

Reply:
I wanted to introduce things gradually, so they are easier to review.
In the end the Trace could become the Call class.

Reply:
What is the problem with modifying PiCall class right now? It doesn't seem like the patch will be bigger, and it should be quite local still.
And you don't need to replace all PI_CALL to PI_TRACE.

Reply:
The Trace is very different from PiCall, and I wouldn't want to mix them.
Notice, even the call-side is different.

Reply:
So, now I'm confused. When I want to call some pi function which macro should I use? PI_CALL or PI_TRACE?

Reply:
Any one today, but PI_TRACE in the long run (I may rename it to PI_CALL when all the uses are changed).

Reply:
Still don't understand the idea. Tracing is supposed to work for all calls, it's misleading otherwise.
Why did you replace only one call site with PI_TRACE? It would be logical to change all call sites or none. It seems now piDeviceGet will have different logging from calls that do PI_CALL.

Reply:
I am trying to go with smaller steps for the benefit of reviewers.
Of course, the PI_TRACE will become the only tracing mechanism with the subsequent changes.

Reply:
I'm OK with all the code except changing PI_CALL to PI_TRACE for getdevice API.

Reply:
> PI_TRACE(RT::piDevicesGet)(...

This is the first use of the new interface, and it is serving the illustration purpose, and for the sanity testing too. I will change the rest of PI_CALLs to PI_TRACE shortly after.

Reply:
Ok, hope we will not have problems before full update.